### PR TITLE
Bitswap: move providing -> Exchange-layer, providerQueryManager -> routing

### LINKED ISSUE #640: Remove requirement on ProviderQueryManager from the bitswap client
## Background

This is the ProviderQueryManager that is in an internal package of the bitswap client

https://github.com/ipfs/boxo/blob/2816b715439a3472ccf7d78e04d5db3b22952728/bitswap/client/internal/providerquerymanager/providerquerymanager.go#L71-L78

Notably it does almost nothing that's specific to Bitswap, or even the concept of data transfer sessions. Instead it is an opinionated content routing wrapper that does things like:
1. Have max timeouts
2. Have predetermined max provider records to use
3. Does preconnects before returning the peers
4. Deduplicates ongoing queries
5. Rate limits requests

Unfortunately sometimes these opinions are too strong for consumers who'd like to experiment or use something else (e.g. higher rate limits, different timeouts, etc.

## Proposal

- Move the ProviderQueryManager package into the routing package
- Refactor it to match the `FindProvidersAsync` method from go-libp2p's content routing interface
- Continue using it by default in the bitswap client, but have an option to not use it
- Add some cleanup / options to the ProviderQueryManager for now and consider separating out the pieces later

Some advantages of this proposal are:
1. If these wrappers are needed/useful elsewhere they can be
2. While we can for now leave the ProviderQueryManager used by the Bitswap client by default we can add an option to disable it and just use the ContentRouting directly
3. While doing 2 we can allow the user to easily provide a modified ProviderQueryManager if it's useful to them by just passing it in as the ContentRouter.

## Alternatives Considered

- Do experiments in forks for the best configuration and upstream to everyone in bitswap/client -> Too cumbersome and inflexible
- Extract into `routing` but disable by default in the bitswap client -> could potentially cause problems for people updating without reading release notes, but would be much cleaner since then we could just leave this as a content routing wrapper
- Extract the ProviderQueryManager into multiple different components for each of the jobs it does -> Might be a good idea, but can also happen later. Particularly if we make the changes in the near future any breakages people experience won't be significant
- Extract into https://github.com/libp2p/go-libp2p-routing-helpers -> I sort of wish this package was already moved into boxo (maybe it should be) and at the moment I'd rather not add more things to go-libp2p-routing-helpers that could reasonably fit in boxo unless downstreams want it to be separate (if you're a downstream and you care, here's where to let me know I'm wrong 😉)

### PR DESCRIPTION
**Pull request by aschmahmann**
fixes #640

This is an attempt at splitting the ProviderQueryManager out of Bitswap so that we can configure it more in consumers.

### CODE REVIEW

File: bitswap/client/internal/session/session.go
Comment (aschmahmann) [inline bitswap/client/internal/session/session.go]: If we're going to keep the function here as not taking a `count` the name should get changed from `FindProvidersAsync` to one that doesn't collide with the more widely used function that does take a count so implementations can use both if they want.

File: bitswap/client/internal/session/session.go
Comment (aschmahmann) [inline bitswap/client/internal/session/session.go]: Note: this was copied out of the ProviderQueryManager because the internals (or at least the tests) seem to rely on having the connections made explicit here. It seems reasonable to extract this up a level out of the routing component though.

**Comment by gammazero** [comment]
@aschmahmann Does this mean we still want the provider query manager, and need to close #536 which removes it.

File: bitswap/options.go
Comment (hsanjuan) [inline bitswap/options.go:112]: ?

File: bitswap/client/client.go
Comment (hsanjuan) [inline bitswap/client/client.go]: In #535 , the Network is separated from the Router and that makes sense to me. As a user, if I want to use a custom Router, I am going to have to glue the normal Network (ConnectTo) and my custom Router (FindProvidersAsync) into a custom BitswapNetwork. I'm not sure if that requires much more refactoring though, given that #535 is based on other additional changes.

This would also allow to disable the ProviderQueryManager altogether when it is not passed in (here it's impossible).

**Review by hsanjuan — APPROVED** [review]
I guess this comes with the need not to replace the default provider itself, but to be able to create it by hand and turn the knobs (and guessing it's mostly about increasing the amount of workers).

It is similar to #535 although that includes some other things (FindProvidersAsync returns AddrInfos instead of PeerIDs, and Router and Network are separate things).

If we need it we can merge this, otherwise the direction of #535 seems better but I guess it's a more invasive change (haven't followed too much). Code-wise I don't see anything weird, so green light on that front.

File: bitswap/options.go
Comment (aschmahmann) [reply bitswap/options.go:112]: This was an unrelated change and should be a separate PR, I'll raise it separately. TLDR though is that `bitswap.New` can take options for the client or the server, but right now the only way to construct them is if they're forwarded as regular Bitswap options, otherwise you have to construct the client and server yourself. This saves on duplication while being more explicit and was something I ran into here because a new client option was introduced.

File: bitswap/network/ipfs_impl.go
Comment (hsanjuan) [inline bitswap/network/ipfs_impl.go:392]: All of this is not needed because our query manager is doing `Connect(peerAddr)` instead of `ConnectTo(peerID)`. Addresses will be absorbed into the peerstore directly. At his point BSnet is only a convenience to have content router available where it is not passed explicitally.

File: routing/providerquerymanager/providerquerymanager.go
Comment (hsanjuan) [inline routing/providerquerymanager/providerquerymanager.go:341]: The beauty is this can now be a libp2p.Host directly. No longer need to rely on the network having absorbed provider peer addresses into the peerstore prior to connecting because we have them here already.

File: bitswap/client/client.go
Comment (hsanjuan) [inline bitswap/client/client.go]: If we don't want to pass-in a Host and a ContentRouter directly to the Client, we need to rely on Network, which has them, but it would be better to remove Content Routing capabilities from the Network because it is just wrapping the content Router and nothing else at this point.

**Review by hsanjuan — DISMISSED** [review]

**Comment by hsanjuan** [comment]
lol I meant to just comment, not approve my own changes

File: bitswap/client/client.go
Comment (aschmahmann) [inline bitswap/client/client.go]: Should this take a function / interface instead of a ProviderQueryManager struct? There are very few exposed functions on the struct (IIUC only FindProvidersAsync / `ProviderQueryRouter` is actually used)

File: routing/mock/centralized_client.go
Comment (aschmahmann) [inline routing/mock/centralized_client.go:54]: This is an unrelated fix due to this code not understanding that 0 means infinity, right?

File: bitswap/network/ipfs_impl.go
Comment (aschmahmann) [inline bitswap/network/ipfs_impl.go:383]: Do we still need to filter this out, or is that not necessary?

**Review by aschmahmann — COMMENTED** [review]
Thanks @hsanjuan for the changes. Looks reasonable to me, left a couple questions

File: bitswap/client/client.go
Comment (aschmahmann) [reply bitswap/client/client.go]: If we make this a function that takes a bsnet and spits out a ProverQueryRouter would that reduce the likelihood of people making mistakes from having two routing systems (one unused in the bsnet and one here).

File: routing/mock/centralized_client.go
Comment (hsanjuan) [reply routing/mock/centralized_client.go:54]: I don't know, you did this change :man_shrugging:

**Comment by hsanjuan** [comment]
Sync with @aschmahmann :

- ProviderQueryManager looks a lot like a ContentRouter itself.
- Hector proposes removing the ContentRouter (CR) part from BitSwapNetwork and passing it directly to Bitswap.New(). This is a breaking change (people need to pass CR to Bitswap.New() rather than to the bsnet.New().
- Adin wants to make ContentRouter completely optional in Bitswap Client, after removing from Server. That provides further justification to break bsnet and Bitswap.New() behaviour. 
- When ContentRouter is provided, it might already be a disguised ProviderQueryManager, so users could:
  - Use no ContentRouting
  - Use their custom ContentRouting
  - Use ProviderQueryManager
  - This means users would need to remember to manually create ProviderQueryManager on upgrade, to keep things working as before. That's shitty and they will likely forget.
  - Alternatively bitswap can:
    - Receive a ContentRouting
    - Also have an option about whether that content routing should be wrapped in a default ProviderQueryManager or not. Defaulting to yes.

I think this is the gist @aschmahmann . I think the last option would be ok (there's breakage but it both a) has a stronger reason b) is easy to fix.

File: routing/mock/centralized_client.go
Comment (aschmahmann) [reply routing/mock/centralized_client.go:54]: oh, lol. Yeah, that's why it's there. I was noticing some weirdness with testing.

**Comment by aschmahmann** [comment]
> I think the last option would be ok (there's breakage but it both a) has a stronger reason b) is easy to fix.

Do you mean to do this?
1. Remove the Providing out of the Bitswap server (e.g. something like https://github.com/ipfs/boxo/pull/677)
2. Break the `bitswap.Network` by removing the ContentRouting
3. Add two options to bitswap.Client, 1 for adding a ProviderFinder and another for whether the default ProviderQueryManager should be used

Existing consumers would migrate by moving their ContentRouter from bitswap.Network into the bitswap.Client option

If so, that sounds reasonable to me. If you think pulling out the Providing is too much to do here we could also do something like:
1. Break the `bitswap.Network` by removing the ContentRouting
2. Add two options to bitswap.Bitswap, 1 for adding a ContentRouting and another (in bitswap.Client) for whether the default ProviderQueryManager should be used
...
In a follow up PR:
1. Remove the Providing out of the Bitswap server
2. Change the bitswap.Bitswap option to forward to a bitswap.Client option and pass a ProviderFinder instead of a ContentRouter (is just a subset so nothing should break)

**Comment by hsanjuan** [comment]
> 1. Remove the Providing out of the Bitswap server (e.g. something like https://github.com/ipfs/boxo/pull/677)
> 2. Break the bitswap.Network by removing the ContentRouting
> 3. Add two options to bitswap.Client, 1 for adding a ProviderFinder and another for whether the default ProviderQueryManager should be used


ok, currently we are here... see my review for a couple of things

File: bitswap/client/client.go
Comment (hsanjuan) [inline bitswap/client/client.go]: I find this problematic as it will silently disable providerQuery-ing if users do not remember to pass in the new option with a content router.

I would prefer to break New() and require a new router there (supporting that it is nil).

File: bitswap/client/client.go
Comment (hsanjuan) [inline bitswap/client/client.go]: We could have a `ProviderFinder` interface here instead of Routing. I think users may not necessarily figure out that a ProviderFinder is a subset of their already existing content routing though, so it is more self-documenting this way.

File: bitswap/testnet/virtual.go
Comment (hsanjuan) [inline bitswap/testnet/virtual.go]: This can be removed now, but more important, we need to wire the mock content-router to the tests at a higher level (instance generator etc) as it is no longer in the network. currently the tests work without content routing.

File: bitswap/client/client.go
Comment (aschmahmann) [reply bitswap/client/client.go]: > I would prefer to break New() and require a new router there (supporting that it is nil).

That's fine by me. In general I prefer options to passing nil's but for new consumers (as opposed to ones who will migrate during the breaking change to bsnet) I can see the value in making them think about whether they want to provide a content router or not.

File: bitswap/client/client.go
Comment (aschmahmann) [reply bitswap/client/client.go]: I find the coupling of these unfortunate and non-Golike as it can end up resulting in wrapper layers and such. It's a bit like how the libp2p host is passed around everywhere even when only a fraction of its functionality is needed. Would just documenting this function and calling out that a `routing.ContentRouting` (https://github.com/libp2p/go-libp2p/blob/dc97b73e4d2970142e366e047a34853a5f54fd8c/core/routing/routing.go#L26) also suffices.

Note: even routing.Routing is way bigger than what we need. If breaking apart the interface in go-libp2p/core/routing would be helpful I'd be that'd be reasonable to the go-libp2p maintainers.

Comment (hsanjuan) [reply bitswap/client/client.go]: Let's do that... I will open a PR on libp2p to split it and we can swap later.

Comment (hsanjuan) [reply bitswap/client/client.go]: https://github.com/libp2p/go-libp2p/pull/3048

File: bitswap/network/ipfs_impl.go
Comment (hsanjuan) [reply bitswap/network/ipfs_impl.go:383]: This is handled now... libp2p would return ErrDialToSelf so we need to ignore that explicitally as we don't know who we are and only have a general purpose dialer.

File: CHANGELOG.md
Comment (hsanjuan) [inline CHANGELOG.md:166]: Reminder to fix this before merge

File: provider/provider.go
Comment (hsanjuan) [inline provider/provider.go:21]: I have updated this interface so that it matches the signature from ContentRouting, so that we can drop this interface altogether if they merge https://github.com/libp2p/go-libp2p/pull/3048

File: bitswap/testinstance/testinstance.go
Comment (hsanjuan) [inline bitswap/testinstance/testinstance.go:24]: We need to pass it in from tests because some tests do want to share a mock router among the different things that use it (Blockservice/Instance etc).

File: bitswap/testinstance/testinstance.go
Comment (hsanjuan) [inline bitswap/testinstance/testinstance.go:91]: There is a test that needs the identity to create a mock client provider on a different provider server to test that blocks provided on a separate network are not received etc etc. so this was necessary.

File: bitswap/client/bitswap_with_sessions_test.go
Comment (hsanjuan) [inline bitswap/client/bitswap_with_sessions_test.go:45]: Since not done by network, it needs to be done here explicitally now.

File: bitswap/client/client.go
Comment (hsanjuan) [inline bitswap/client/client.go:208]: Allowing to pass nil parameters is not awesome either...

File: bitswap/client/internal/session/session.go
Comment (hsanjuan) [inline bitswap/client/internal/session/session.go]: To discuss: make this `FindProvidersAsync(ctx context.Context, k cid.Cid, max int)`. Consider this a `libp2p.ContentDiscovery` interface and not a custom one. Let the ProviderQueryManager implement that. Let `max` be a bitswap.Option and not a ProviderQueryManager one, or call it with `-1` or something to signal we want to use the "default".

File: blockservice/blockservice.go
Comment (hsanjuan) [inline blockservice/blockservice.go]: Confirm we are definitely removing this

File: blockservice/providing_blockstore.go
Comment (hsanjuan) [inline blockservice/providing_blockstore.go]: It was noted that this assumes the Provider does things async etc... is this the case?

File: blockservice/blockservice.go
Comment (hsanjuan) [reply blockservice/blockservice.go]: Decision: no, we don't want to remove this. Leave it

File: blockservice/providing_blockstore.go
Comment (hsanjuan) [reply blockservice/providing_blockstore.go]: Decision: port the old logic. Make this part of an Exchange implementation (ProvidingExchange).

File: bitswap/client/internal/session/session.go
Comment (hsanjuan) [reply bitswap/client/internal/session/session.go]: @aschmahmann @gammazero

File: provider/provider.go
Comment (hsanjuan) [reply provider/provider.go:21]: It was merged, now wait for libp2p release

File: blockservice/blockservice.go
Comment (hsanjuan) [reply blockservice/blockservice.go]: These changes are now undone.

**Comment by hsanjuan** [comment]
@aschmahmann @gammazero I have a question: I was going to port the "Provider" logic from bitswap server, with its workers and queues and what not. Then I saw that we have a provider.System (the reprovider) that does have queues, does batching and even stores pending provides to disk. Naively my thought is that, if such logic is already implemented in the reprovider, a providing.Exchange should not yet have its own way and we should just pass-in the reprovider when using it. Plus I suspect that we make use of things like Accelerated-DHT-providing, `ProvideMany`, so why going around...

What do you think?

File: blockservice/providing_blockstore.go
Comment (hsanjuan) [reply blockservice/providing_blockstore.go]: See comment below: https://github.com/ipfs/boxo/pull/641#issuecomment-2489038266

**Comment by aschmahmann** [comment]
>What do you think?

Makes sense to me to leave it out. It's unclear how useful the parallel queues were beforehand, but it seems like the kind of thing that's situational enough that people can choose to add whatever queues they need and we should start off keeping it simple and encouraging reuse of the existing queue.

File: bitswap/client/internal/session/session.go
Comment (aschmahmann) [reply bitswap/client/internal/session/session.go]: I think using the `libp2p.ContentDiscovery` interface (i.e. with the `max int`) makes sense.

> Let max be a bitswap.Option and not a ProviderQueryManager one, or call it with -1 or something to signal we want to use the "default".

Why not have bitswap call it with 0 (i.e. give me all of them) and have the ProviderQueryManager handle any "max" setting? The "max" on the FindProvidersAsync interface is about the max providers returned rather than the max providers searched for. The ProviderQueryManager could reasonably choose to search for 100 peers but only return the first 10 that it connected to and still be honoring `FindProvidersAsync(..., 10)`, even though right now it will search for 10 and return <= 10 even if there are 1000 providers that exist and 10 that are reachable.

File: blockservice/providing_blockstore.go
Comment (hsanjuan) [reply blockservice/providing_blockstore.go]: Resolved by leaving a simple wrapper and we will try to use the Reprovider directly for this.

File: bitswap/client/internal/session/session.go
Comment (hsanjuan) [reply bitswap/client/internal/session/session.go]: If ProviderQueryManager implements the ContentDiscovery interface, then it should return an unlimited number of results when called with max == 0, per the requirements of that interface.

We could use <0 as the values that we use the MaxProviders option for (the routing interface does not specify what to do with negative values):

  * `pqm.FindProvidersAsync(ctx, cid, 10)`  -> returns at most 10 providers that we manage to connect to (it may search more)
  * `pqm.FindProvidersAsync(ctx, cid, 0)` -> returns unbounded number of providers that we manage to connect to.
  * `pqm.FindProvidersAsync(ctx, cid, -1)` -> returns at most pqm.maxProviders that we manage to connect to.

Thoughts?

File: routing/providerquerymanager/providerquerymanager.go
Comment (hsanjuan) [inline routing/providerquerymanager/providerquerymanager.go:21]: Are we sure we want 0 now?

File: bitswap/client/internal/session/session.go
Comment (hsanjuan) [inline bitswap/client/internal/session/session.go]: It is a bit ugly that here it is assumed that the ContentDiscovery that we use understands what -1 means.

File: bitswap/client/internal/session/session.go
Comment (aschmahmann) [reply bitswap/client/internal/session/session.go]: We could do this, although is it worth the hassle given at the moment there's only one consumer that likely doesn't need this?

What about this?
- `pqm.FindProvidersAsync(ctx, cid, 10)` returns at most 10 providers that we manage to connect to (it will search for up to pqm.maxProviders which can be set to 0 and ignored)
- `pqm.FindProvidersAsync(ctx, cid, 0)` -> returns unbounded number of providers that we manage to connect to (it will search for up to pqm.maxProviders which can be set to 0 and ignored)
- default pqm.maxProviders is 0

In Bitswap we would default to `pqm.FindProvidersAsync(ctx, cid, 10)` and would set `pqm.maxProviders` to 10. In a followup PR (once we know this behaves well) we'd try setting pqm.maxProviders to 0 by default.

If in the future there's a desire to effectively set `pqm.maxProviders` per query as a parameter we can figure something out like:
1. Just add a new function `FindReachableProvidersAsync(ctx, cid, maxReturnedProviders, maxAttemptedProviders)`
2. Hide the `maxProviders` as a context settable option
3. Use negative numbers
   - `pqm.FindProvidersAsync(ctx, cid, -X)` returns at most X providers that we manage to connect to (it will search for up to X providers). We could then drop the ability to set `pqm.maxProviders` entirely and assert that setting the last parameter >= 0 is the same as the old `pqm.maxProviders=0` (i.e. the default)

Comment (hsanjuan) [reply bitswap/client/internal/session/session.go]: yeah ok, then we do your proposal

File: routing/providerquerymanager/providerquerymanager.go
Comment (hsanjuan) [reply routing/providerquerymanager/providerquerymanager.go:21]: Apparently yes

File: bitswap/client/internal/session/session.go
Comment (hsanjuan) [reply bitswap/client/internal/session/session.go]: ok, now calling it with 0.

File: bitswap/client/internal/session/session.go
Comment (hsanjuan) [reply bitswap/client/internal/session/session.go]: fixed in latest commit

**Comment by hsanjuan** [comment]
https://github.com/ipfs/kubo/pull/10595 is testing this.

File: blockservice/providing_blockstore.go
Comment (hsanjuan) [inline blockservice/providing_blockstore.go]: I need to remove this, the blockstore does not need to provide

File: gateway/backend_blocks.go
Comment (hsanjuan) [inline gateway/backend_blocks.go]: Need to re-add this

**Comment by hsanjuan** [comment]
Squashed

**Review by gammazero — APPROVED** [review]

**Pull request merged**
